plugins/services/src/js/pod-instances/PodInstancesContainer.js

```diff
@@ -0,0 +1,227 @@
+
+import {
+  REQUEST_MARATHON_POD_INSTANCE_KILL_ERROR,
+  REQUEST_MARATHON_POD_INSTANCE_KILL_SUCCESS
+} from '../constants/ActionTypes';
+
+const METHODS_TO_BIND = [
+  'handleMesosStateChange',
+  'handleServerAction',
+  'handleModalClose',
+  'clearActionError',
+  'killPodInstances'
+];
+
+class PodInstancesContainer extends React.Component {
+  constructor() {
+    super(...arguments);
+
+    this.state = {
+      actionErrors: {},
+      lastUpdate: 0,
+      pendingActions: {}
+    };
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  componentDidMount() {
+    MesosStateStore.addChangeListener(
+      EventTypes.MESOS_STATE_CHANGE,
+      this.handleMesosStateChange
+    );
+    // Listen for server actions so we can update state immediately
+    // on the completion of an API request.
+    this.dispatcher = AppDispatcher.register(this.handleServerAction);
+  }
+
+  componentWillUnmount() {
+    MesosStateStore.removeChangeListener(
+      EventTypes.MESOS_STATE_CHANGE,
+      this.handleMesosStateChange
+    );
+
+    AppDispatcher.unregister(this.dispatcher);
+  }
+
+  getChildContext() {
+    return {
+      modalHandlers: this.getModalHandlers()
+    };
+  }
+
+  killPodInstances() {
+    this.setPendingAction(ActionKeys.POD_INSTANCES_KILL);
+
+    return MarathonActions.killPodInstances(...arguments);
+  }
+
+  handleMesosStateChange() {
+    this.setState({
+      lastUpdate: Date.now()
+    });
+  }
+
+  handleServerAction(payload) {
+    const {action} = payload;
+
+    switch (action.type) {
+      case REQUEST_MARATHON_POD_INSTANCE_KILL_ERROR:
+        this.unsetPendingAction(ActionKeys.POD_INSTANCES_KILL, action.data);
+        break;
+      case REQUEST_MARATHON_POD_INSTANCE_KILL_SUCCESS:
+        this.unsetPendingAction(ActionKeys.POD_INSTANCES_KILL);
+        break;
+    }
+  }
+
+  handleModalClose(key) {
+    if (key) {
+      this.clearActionError(key);
+    }
+    this.setState({modal: {}});
+  }
+
+  fetchData() {
+    // Re-fetch data - this will end up being a single Relay request
+  }
+  /**
+   * Sets the actionType to pending in state which will in turn be pushed
+   * to children components as a prop. Also clears any existing error for
+   * the actionType
+   * @param {String} actionType
+   */
+  setPendingAction(actionType) {
+    const {actionErrors, pendingActions} = this.state;
+
+    this.setState({
+      actionErrors: ContainerUtil.adjustActionErrors(
+        actionErrors, actionType, null
+      ),
+      pendingActions: ContainerUtil.adjustPendingActions(
+        pendingActions, actionType, true
+      )
+    });
+  }
+  /**
+   * Sets the pending actionType to false in state which will in turn be
+   * pushed down to children via props. Can optional set an error for the
+   * actionType
+   * @param  {String} actionType
+   * @param  {Any} error
+   */
+  unsetPendingAction(actionType, error = null) {
```

Comment:
Let's move all these methods to handle modals into some kind of a modal service instead of passing these actions down using context and parameter. The `PodInstancesContainer` shouldn't need to handle all of this.